### PR TITLE
[TAO-1796][Feature] video_clip: add InternVideo2-CLIP TensorRT deploy module

### PR DESCRIPTION
**Pull request by lianqiann**
## Summary
Adds the deploy-side counterpart of the tao-pytorch `video_clip` task: build a TensorRT engine from the exported InternVideo2-CLIP ONNX, then run retrieval inference and evaluation against it.

## Changes
- **`video_clip` entrypoint** with `gen_trt_engine`, `inference` and `evaluate` subtasks, the shipped experiment spec, and the config dataclasses under `config/multimodal/video_clip/`.
- **Dataloader** mirroring the training-time contract — frame sampling, ImageNet (not CLIP) normalization, and optional text canonicalization gated on the exported model config — so deploy-time preprocessing matches training exactly. A mismatch here degrades retrieval silently rather than failing.
- **Retrieval evaluation** — mAP, Recall@K, median and mean rank over the combined video and text embeddings.
- **`EngineBuilder._extra_network_flags()`** — a new overridable hook on the shared builder, defaulting to `0`, OR-ed into the `create_network` flags. `video_clip` overrides it to request a strongly-typed network, which is what stops TensorRT's weakly-typed fp16 path from re-interpreting the AutoCast mixed-precision ONNX. No behaviour change for any existing network.

## Testing
- Unit tests covering the config, dataloader preprocessing, engine builder, entrypoint wiring, retrieval metrics and inferencer.
- `pre-commit` (docs supported-commands, license header, pylint, pydocstyle, flake8) clean over the full diff range.
- End-to-end engine build, inference and evaluation validated against the exported InternVideo2-CLIP ONNX.

## Migration provenance
Manual migration of GitLab MR [!205](https://gitlab-master.nvidia.com/nvidia-tao-toolkit/tao-deploy/-/merge_requests/205) (branch `alicli/iv2clip`). The GitHub and GitLab repositories share no git history, so the branch's net diff was replayed onto `main` as three logical commits. Two notes:

- The source branch had advanced past the MR snapshot — it now includes the strongly-typed TRT build and the fp16 normalization fix — so the **current branch tip** was ported, not the older MR diff.
- `docs/supported_commands.md` is regenerated by the repo's own `update-docs-supported-commands` hook and is included here.

Depends on the tao-pytorch PR for the exported model this module consumes.

### CODE REVIEW

**Comment by lianqiann** [comment]
Migrated from GitLab as part of the TAO GitHub-first cutover. This change lands across three repositories — the set is:

- tao-pytorch: https://github.com/NVIDIA-TAO/tao-pytorch/pull/94 — the `video_clip` module (GitLab !620)
- tao-deploy: https://github.com/NVIDIA-TAO/tao-deploy/pull/36 — the TensorRT deploy module (GitLab !205)
- tao-core: https://github.com/NVIDIA-TAO/tao-core/pull/30 — the config dataclasses and schema

The tao-pytorch PR is self-contained and can merge independently: the module does not import `nvidia_tao_core`. tao-deploy consumes the ONNX that tao-pytorch exports, and tao-core only brings the API/microservices schema into parity.

File: nvidia_tao_deploy/multimodal/video_clip/evaluation/retrieval.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+                ranks.append(first_pos_rank[0] + 1)
+
+            for k in self.k_values:
+                hits_at_k = np.sum(sorted_labels[:k])
+                recall_scores[k].append(hits_at_k / n_pos)
+                hit_scores[k].append(1.0 if hits_at_k > 0 else 0.0)
+                ndcg_scores[k].append(compute_ndcg(sorted_labels, k))
+
+            if self._compute_auc:
+                auc_scores.append(compute_auc(sims, sorted_labels))
```
Comment (vpraveen-nv) [inline nvidia_tao_deploy/multimodal/video_clip/evaluation/retrieval.py]: Bug: `sims` is in gallery order but `sorted_labels` is in descending-rank order, so `compute_auc` pairs each gallery item's score with the relevance of the wrong item (it sorts `scores` internally and indexes `labels` with the result, so the two arrays must be aligned). Should be `compute_auc(sims[sorted_idx], sorted_labels)` (or pass gallery-order labels). AUC only rides along in the JSON report, but as reported it's meaningless. This is copied verbatim from the tao-pytorch twin (`model/evaluation/retrieval.py` L245), so the fix should land in PR #94 too.

File: nvidia_tao_deploy/multimodal/video_clip/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    video_path = str(video_path)
+    mapping = path_prefix_mapping or {}
+
+    for prefix, replacement in mapping.items():
+        if prefix and video_path.startswith(prefix):
+            return video_path.replace(prefix, str(replacement), 1)
+
+    path = Path(video_path)
+    if path.is_absolute():
+        if data_root and video_path.startswith("/media/wbf/"):
```
Comment (vpraveen-nv) [inline nvidia_tao_deploy/multimodal/video_clip/dataloader.py]: Same `/media/wbf/` hardcoded remap as the trainer (flagged on tao-pytorch PR #94) — since this is intentionally a verbatim port, whatever we decide there (drop the special case in favor of `path_prefix_mapping`) should land here in the same pass to keep the two copies identical.

File: nvidia_tao_deploy/multimodal/video_clip/scripts/evaluate.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+def load_model_config(trt_engine_path: str) -> dict:
+    """Load model_type / canonicalize_text / tokenizer_path from export artifacts."""
+    artifacts = find_export_artifacts(trt_engine_path)
+    if artifacts['tokenizer_path'] is None:
+        raise FileNotFoundError(
+            f"No *_tokenizer/ directory found for engine {trt_engine_path}. "
+            "Copy the tokenizer from ONNX export next to the engine."
+        )
+    model_type, canonicalize = "internvideo2-clip-l14", False
```
Comment (vpraveen-nv) [inline nvidia_tao_deploy/multimodal/video_clip/scripts/evaluate.py:81]: When `*_config.yaml` isn't found, `canonicalize_text` silently defaults to False — which is exactly the silent-accuracy-degradation mode the dataloader docstring warns about (a model exported with `canonicalize_text: true` would evaluate with mismatched tokens and just score worse). Since the tokenizer is already required, suggest at least a loud warning (or an error) when the config yaml is missing rather than a silent default.

File: nvidia_tao_deploy/config/multimodal/video_clip/default_config.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                    "chunk_id to its video path and time window.",
+        display_name="Chunk Metadata",
+    )
+    video_root: Optional[str] = STR_FIELD(
+        value=None,
+        default_value=None,
+        description="Optional root prepended to relative video paths.",
+        display_name="Video Root",
+    )
+    path_prefix_mapping: Dict[str, str] = field(default_factory=dict)
```
Comment (vpraveen-nv) [inline nvidia_tao_deploy/config/multimodal/video_clip/default_config.py]: Nit: plain `field(default_factory=dict)` while the rest of the schema (and the tao-pytorch config, which uses `DICT_FIELD` for this same key) goes through the typed field helpers — worth switching to `DICT_FIELD` so the description/display metadata shows up in default_specs like every other field.

**Review by vpraveen-nv — COMMENTED** [review]
Thanks Alice — clean deploy module, and the preprocessing-contract documentation in the dataloader is exactly the right emphasis. I checked GitLab !205 (no prior review comments there) so nothing here is a repeat. One real bug in the AUC computation (inherited from the tao-pytorch twin — worth fixing in both), plus a few smaller items inline.

File: nvidia_tao_deploy/multimodal/video_clip/evaluation/retrieval.py (hunk 1 above)
Comment (lianqiann) [reply nvidia_tao_deploy/multimodal/video_clip/evaluation/retrieval.py]: Fixed, and the bug was worse than "meaningless" — the reported AUC was **inverted**.

Now `compute_auc(sims[sorted_idx], sorted_labels)`.

I also went looking for why this shipped: the AUC path had **zero** test coverage. Every existing case constructed the evaluator with `compute_auc=False` (`tests/video_clip/test_evaluation.py:32` and `:66`), so nothing ever executed line 136. Added `TestRetrievalAUC` with a gallery whose ranking order (`[3,2,1,0]`) deliberately differs from gallery order, so a misaligned call cannot coincidentally agree.

Verified the tests are real discriminators by re-running them against the old line:

| case | expected | buggy line | fixed |
|---|---|---|---|
| only lowest-scoring clip relevant | 0.0 | **1.0** | 0.0 |
| only highest-scoring clip relevant | 1.0 | **0.0** | 1.0 |

Mirrored into tao-pytorch PR #94 as you asked — same one-line fix at `model/evaluation/retrieval.py:245` plus the equivalent tests.

File: nvidia_tao_deploy/multimodal/video_clip/dataloader.py (hunk 2 above)
Comment (lianqiann) [reply nvidia_tao_deploy/multimodal/video_clip/dataloader.py]: Removed here too, matching what landed on #94 — `_resolve_video_path` is now just `if path.is_absolute(): return str(path)`, byte-identical to `video_text_loader.py:134-139` on the tao-pytorch branch. `path_prefix_mapping` is checked first and covers the case.

No test churn: nothing baked in a `/media/wbf/` path, and `test_path_prefix_mapping` already covered the replacement mechanism. Added `test_data_root_does_not_rewrite_absolute_paths` to pin the new contract — `video_root`/`data_root` prefixes **relative** paths only, remapping absolute ones is `path_prefix_mapping`'s job.

File: nvidia_tao_deploy/multimodal/video_clip/scripts/evaluate.py (hunk 3 above)
Comment (lianqiann) [reply nvidia_tao_deploy/multimodal/video_clip/scripts/evaluate.py:81]: Added a loud warning naming the engine, the directories searched, and both assumed values.

I kept it non-fatal rather than raising, because I traced how far the flag actually reaches and a hard error would block evaluation over something that cannot currently bite:

- `model.canonicalize_text` is optional with a `False` default in both repos (`tao-pytorch config/video_clip/default_config.py:110`, `tao-deploy config/multimodal/video_clip/default_config.py:44`).
- `export.py:1058-1061` writes `<base>_config.yaml` **unconditionally** beside the ONNX and the `_tokenizer/` dir, so it only goes missing when someone hand-copies engine + tokenizer and drops the yaml.
- More to the point: **InternVideo2-CLIP ignores the flag at train time.** `build_internvideo2clip_model` (`builders.py:402`) never passes it, `InternVideo2Tokenizer` (`adapters/internvideo2clip.py:43-59`) never applies it, and `pl_video_clip_model.py` has no references — only the SigLIP2/OpenCLIP/RADIO tokenizer wrappers honor it. So for the one model type this deploy module supports, `False` is the provably matching value, not just the default.

The warning documents the assumption for the day another backbone lands here. Worth flagging separately: the inverse is a live footgun upstream — setting `canonicalize_text: true` in an IV2CLIP training spec is silently ignored during training but *is* honored by `evaluate.py:108`, which would create exactly the skew you describe in the opposite direction. That is a tao-pytorch issue; happy to open it if you agree.

File: nvidia_tao_deploy/config/multimodal/video_clip/default_config.py (hunk 4 above)
Comment (lianqiann) [reply nvidia_tao_deploy/config/multimodal/video_clip/default_config.py]: Switched to `DICT_FIELD` (`config/utils/types.py:287`) with description and display name, matching the tao-pytorch config for the same key (`config/video_clip/default_config.py:418-422`). It was the only key in this schema bypassing the typed helpers. Dropped the now-unused `field` import from `dataclasses`.

**Review by vpraveen-nv — APPROVED** [review]

**Review by christinayyw — APPROVED** [review]
